subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/result/DefaultResolvedArtifactResult.java

```diff
@@ -24,14 +24,20 @@
 import java.io.File;
 
 public class DefaultResolvedArtifactResult implements ResolvedArtifactResult {
+    private static final String UNDEFINED_VARIANT_NAME = "undefined";
+
     private final ComponentArtifactIdentifier identifier;
     private final ResolvedVariantResult variant;
     private final Class<? extends Artifact> type;
     private final File file;
 
     public DefaultResolvedArtifactResult(ComponentArtifactIdentifier identifier, AttributeContainer variantAttributes, Class<? extends Artifact> type, File file) {
+        this(identifier, UNDEFINED_VARIANT_NAME, variantAttributes, type, file);
```

Comment:
This isn't right. If the caller provides the variant attributes, they should also provide the variant display name.
It looks like `ArtifactVisitor.visitArtifact()` and `visitFile` could either take a `ResolvedVariant` in place of `AttributeContainer` or have an additional 'variantName' parameter.

For each caller of `ArtifactVisitor.visitArtifact()` you should be able to provide a reasonable display name. Certainly `ArtifactBackedResolvedVariant` already has a 'displayName' property that might suffice.